xdsl/irdl/constraints.py

```diff
@@ -72,24 +82,23 @@ def verify(
         """
         ...
 
-    def get_resolved_variables(self) -> set[str]:
+    def get_resolvers(self) -> dict[str, Resolver[AttributeCovT]]:
         """
-        Get the set of type variables that are always resolved when verifying
-        the constraint.
+        Get a dictionary of variables that can be solved from this attribute.
         """
-        return set()
+        return dict()
 
-    def can_infer(self, constraint_names: set[str]) -> bool:
+    def can_infer(self, variables: Set[str]) -> bool:
         """
         Check if there is enough information to infer the attribute given the
         constraint variables that are already set.
         """
         # By default, we cannot infer anything.
         return False
```

Comment:
Because we have get_resolvers, this function should always be defined?
Is there a reason we could infer something that doesn't have a resolver, or vice versa?

Reply:
Any `EqAttrConstraint` can be inferred but does not resolve any variables. A `ParamAttrConstraint` may be able to provide variable resolutions without being totally inferrable

Reply:
And in general there are lots of things that cannot be inferred, and must be given in the assembly format

Reply:
Oh yes you're right, my bad.